dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java

```diff
@@ -896,4 +899,271 @@ public VersionTypeValidation validateToolSet(Set<SourceFile> sourcefiles, String
     public VersionTypeValidation validateTestParameterSet(Set<SourceFile> sourceFiles) {
         return checkValidJsonAndYamlFiles(sourceFiles, DescriptorLanguage.FileType.CWL_TEST_JSON);
     }
+
+    /**
+     * Implements a preprocessor which "expands" a CWL, replacing $import, $include, $mixin, and "run" directives per the CWL
+     * spec https://www.commonwl.org/v1.2/Workflow.html, using the content of the referenced source files, with the exception
+     * of a "run" directive that points to a missing source file, which is normalized to the "run: file" syntax and otherwise
+     * left unchanged.
+     *
+     * <p>To facilitate the extraction of information from a CWL that is missing files, if an $import references a missing file,
+     * it is replaced by the empty Map.  If an $include references a missing file, it is replaced by the empty string.
+     *
+     * <p>Typically, a Preprocessor instance is one-time-use: a new Preprocessor instance is created to expand each root CWL
+     * descriptor.
+     *
+     * <p>As the preprocessor expands the CWL, it tracks the current file, and for each entry (workflow or tool) it encounters,
+     * it first ensures that the entry has a unique id (by assigning the missing or duplicate id to a UUID), then adds the
+     * id-to-current-file-path relationship to a Map.  Later, a parser can query the Map via the getPath method to determine
+     * what file the entry came from.
+     *
+     * <p>During expansion, the preprocessor tracks three quantities to prevent denial-of-service attacks or infinite
+     * loops due to a recursive CWL: the file import depth, the (approximate) total length of the expanded CWL in characters, and
+     * total number of files expanded (incremented for each $import, $include, $mixin, and "run" directive).  If any of those
+     * quantities exceed the maximum value, the preprocessor will call the handleMax function, the base implementation of which
+     * will throw an exception.
+     */
+    public static class Preprocessor {
+        private static final List<String> IMPORT_KEYS = Arrays.asList("$import", "import");
+        private static final List<String> INCLUDE_KEYS = Arrays.asList("$include", "include");
+        private static final List<String> MIXIN_KEYS = Arrays.asList("$mixin", "mixin");
+        private static final String EMPTY_STRING = "";
+        private static final Map<String, Object> EMPTY_MAP = Collections.emptyMap();
+        private static final int DEFAULT_MAX_DEPTH = 10;
+        private static final long DEFAULT_MAX_CHAR_COUNT = 4L * 1024L * 1024L;
+        private static final long DEFAULT_MAX_FILE_COUNT = 2000L;
```

Comment:
I hope nobody submits a 2,000 descriptor workflow. :) Does a tad high, but no idea what we should set it to, and unlikely to be run into.

Reply:
wonder if we should make these config limits like our workflow and workflow version limit
https://github.com/dockstore/compose_setup/blob/develop/templates/web.yml.template#L39

Easy to update without a new deploy, but can act as a guard against abuse in the meantime

Reply:
> Haven't looked at the code yet, but questions from skimming your description:
> 
> > $mixin is supported, for all CWL versions. However, it was dropped from the spec a while back. Ok to keep supporting? $mixin/mixin are likely rarely used as normal keys.
> 
> 1. Do we have source files with mixin?
> 2. What version of CWL was it dropped? How hard would it be to honor the version if it was supported in older versions?

I inspected a recent DB dump, and it looks like about one in 100 sourcefiles that end with '.cwl' contain the word 'mixin'.  All of such source files appear to be CWL v1.0.

I will modify so mixins are only preprocessed in v1.0 files.

Reply:
`java/io/dockstore/client/cli/DAGWorkflowTestIT.java` refers to an actual workflow with 229 DAG nodes.  Each node could theoretically have `$include`s, so I'd be worried about setting `MAX_FILE_COUNT` below 1000.

As an aside, a big issue with megaworkflows is that the sourcefile import code is _slow_, it appeared to grab files at a rate of 2-3 per second.  To gracefully handle hundreds of sourcefiles, we'd probably want to thread the retrieval, or, even better, grab the full repo as a single file and pluck the files from it.